build/block.rs

```diff
@@ -522,9 +539,111 @@ pub fn build() -> anyhow::Result<TokenStream> {
                 }
             }
 
+            pub fn block_entity_type(self) -> Option<u8> {
+                match self.0 {
+                    #state_to_block_entity_type_arms
+                    _ => None
+                }
+            }
+
             #default_block_states
         }
 
+        #[derive(Clone, PartialEq, Debug)]
+        pub struct BlockEntity {
+            pub kind: VarInt,
+            pub data: BlockEntityData,
+        }
+
+        impl BlockEntity {
+            pub fn new(data: BlockEntityData) -> Self {
+                Self {
+                    kind: data.get_kind(),
+                    data,
+                }
+            }
+        }
+
+        #[derive(Clone, PartialEq, Debug)]
+        pub enum BlockEntityData {
+            Sign(SignData),
+        }
+
+        impl BlockEntityData {
+            pub fn to_compound(&self) -> anyhow::Result<Compound> {
+                match self {
+                    BlockEntityData::Sign(data) => {
+                        Ok(Compound::from(
+                            [
+                                ("GlowingText".to_string(), Value::Byte(if data.glowing_text {1} else {0})),
+                                ("Color".to_string(), Value::String(data.color.to_string())),
+                                ("Text1".to_string(), Value::String(serde_json::to_string(&data.texts[0])?)),
+                                ("Text2".to_string(), Value::String(serde_json::to_string(&data.texts[1])?)),
+                                ("Text3".to_string(), Value::String(serde_json::to_string(&data.texts[2])?)),
+                                ("Text4".to_string(), Value::String(serde_json::to_string(&data.texts[3])?)),
+                            ]))
+                    }
+                }
+            }
+
+            pub fn get_kind(&self) -> VarInt {
+                match self {
+                    BlockEntityData::Sign(_) => VarInt(7),
+                }
+            }
+        }
+
+        #[derive(Clone, PartialEq, Debug)]
+        pub struct SignData {
+            pub glowing_text: bool,
+            pub color: TextColor,
+            pub texts: [Text; 4],
+        }
+
+        #[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+        pub enum TextColor {
```

Comment:
We already have a `Color` type in `text.rs` -- would it be possible to reuse that here?

Reply:
While they might have the same values, in vanilla minecraft they're from two different sources (doesn't mean we can't use it obviously, just one thing to consider)

Reply:
Also - the `Color` type in `text.rs` consists of rgb values with constants for each color. This is not the behavior we want here, as signs only have that enum. Thus I will keep my enum